### ISSUE TITLE
fix maintainer-script-ignores-errors

### ISSUE THREAD

**Issue by root360-AndreasUlm** [jordansissel/fpm#1696]
Hi,

although 'maintainer-script-ignores-errors' is only a warning in lintian the effect can be vital.
If an important command fails within a bad written script the whole maintainer script can do a lot of harm.

See this sample scenario:

Assume the --after-install script looks like this:
```
mkdir /tmp/test
cd /tmp/test
[generate temporary files and directories]
mv ./* /usr/share/<package>/
rm -rf ./*
```

When the creation of `/tmp/test` works everything is working as intended.
But when the creation fails e.g. because `/tmp/test` is a file the change of directory fails and every following command is run in `/` thus creating random stuff there and at last removing all files and directories.
This can be prevented by setting the error exit flag of the shell.

That is why I'd like the flag to be added to the maintainer scripts.

I'll create a pull request for that soon.

Regards,
Andreas

**Comment by BartKoppers** [comment]
A maintainer could also decide to include set -e in the script itself.

It's a simple check to test if /tmp/test does exist and is a directory. 
If this is important, the shellscript should solve this. And IMHO not fpm.

But, although I think this doesn't really solve these initial problems, it doesn't 'hurt' either.
Well,  *at least where /bin/sh is linked to Bash* (on Mac Catalina it isn't anymore). So you could (should?) check this.

**Comment by root360-AndreasUlm** [comment]
> A maintainer could also decide to include set -e in the script itself.

Yes a maintainer could but he/*/she cannot set it globally as the scripts provided via the cli-parameters are merged into a complete script based on skeletons defined by fpm.
Thus the maintainer can never solve the lintian error [maintainer-script-ignores-errors](https://lintian.debian.org/tags/maintainer-script-ignores-errors.html) on his own.

> It's a simple check to test if /tmp/test does exist and is a directory.
> If this is important, the shellscript should solve this. And IMHO not fpm.

The sample is just a simple one to show the point I'm talking about.

> Well, _at least where /bin/sh is linked to Bash_ (on Mac Catalina it isn't anymore). So you could (should?) check this.

The shell option 'errexit' which is activated when `set -e` is executed is supported by all modern POSIX-compliant shells [1] and is not a Bash-only feature.
Debian uses since a long time Dash as default shell.

[1]:
* zsh https://linux.die.net/man/1/zshoptions (option ERR_EXIT)
* dash https://linux.die.net/man/1/dash (option errexit)
* csh https://linux.die.net/man/1/csh (option -e)
* tcsh https://linux.die.net/man/1/tcsh (option -e)
* ksh https://linux.die.net/man/1/ksh (option -e)
* etc.

**Comment by BartKoppers** [comment]
Hi, thanks and I get the pro's and points from your view. 

BTW, isn't this enhancement solvable through `--template-scripts` (w or w/o `--template-value KEY=VALUE`)?

Did you read [ Why doesn't set -e (or set -o errexit, or trap ERR) do what I expected?](http://mywiki.wooledge.org/BashFAQ/105) ?
I (still) agree with its conclusions: it's difficult to predict if and how - by introduction of `set -e `-  the outcome is changed.
And could introduce a IMHO high price for fixing a Lintian _warning_ (not an error)

Maybe a better option would be introducing a parameter , so `set -e `is only set if stated.
Something along `--deb-maintainerscripts-force-errorchecks`?

> .......
> So-called strict mode
> In the mid 2010s, some people decided that the combination of set -e, set -u and set -o pipefail should be used by default in all new shell scripts. They call this unofficial bash strict mode, and they claim that it "makes many classes of subtle bugs impossible" and that if you follow this policy, you will "spend much less time debugging, and also avoid having unexpected complications in production".
> 
> As we've already seen in the exercises above, these claims are dubious at best. The behavior of set -e is quite unpredictable. If you choose to use it, you will have to be hyper-aware of all the false positives that can cause it to trigger, and work around them by "marking" every line that's allowed to fail with something like ||true.
> 
> **Conclusions**
> GreyCat's personal recommendation is simple: don't use set -e. Add your own error checking instead.
> 
> rking's personal recommendation is to go ahead and use set -e, but beware of possible gotchas. It has useful semantics, so to exclude it from the toolbox is to give into FUD.
> 
> geirha's personal recommendation is to handle errors properly and not rely on the unreliable set -e.

**Comment by root360-AndreasUlm** [comment]
Hi,

I like the idea of providing `--deb-maintainerscripts-force-errorchecks` to optionally activate errexit as it might otherwise break current scripts that do not fully handle errors properly.

I read the link you provided. As I'm mostly writing shell scripts which handle known errors pre-defined and properly I haven't had any issues as described there.
But it's true that enforcing this shell option could lead to issues in script written by poeple who don't have that much shell scripting experience.

I'll change the PR to activate errexit when fpm is called with `--deb-maintainerscripts-force-errorchecks`.
It will be done most likely next week.

**Comment by BartKoppers** [comment]
> 
> I like the idea of providing `--deb-maintainerscripts-force-errorchecks` to optionally activate errexit as it might otherwise break current scripts that do not fully handle errors properly.

Now, I have no idea if @jordansissel has some strict parameter-naming convention in place, but the thought-up `--deb-maintainerscripts-force-errorchecks` seemed fitting in. 
But it is also kinda long :)
Let's say that we're open for suggestions on an even better parameter-name here?

> I'll change the PR to activate errexit when fpm is called with `--deb-maintainerscripts-force-errorchecks`.
> It will be done most likely next week.

Super, Andreas!

**Comment by root360-AndreasUlm** [comment]
@BartKoppers @jordansissel 
implemented errexit fpm parameter --deb-maintainerscripts-force-errorchecks
